File: backend/routes/shipments.js
const express = require('express');
const router = express.Router();
const { v4: uuidv4 } = require('uuid');
const db = require('../db');

router.get('/', (req, res) => {
  const { search, status, driver_id, page = 1, limit = 20 } = req.query;
  let query = `
    SELECT s.*, d.name as driver_name, d.vehicle_plate, o.order_number,
           c.name as customer_name
    FROM shipments s
    LEFT JOIN drivers d ON s.driver_id = d.id
    LEFT JOIN orders o ON s.order_id = o.id
    LEFT JOIN customers c ON o.customer_id = c.id
    WHERE 1=1
  `;
  const params = [];
  if (search) {
    query += ' AND (s.tracking_number LIKE ? OR s.origin LIKE ? OR s.destination LIKE ? OR d.name LIKE ?)';
    params.push(`%${search}%`, `%${search}%`, `%${search}%`, `%${search}%`);
  }
  if (status) { query += ' AND s.status = ?'; params.push(status); }
  if (driver_id) { query += ' AND s.driver_id = ?'; params.push(driver_id); }
  query += ' ORDER BY s.created_at DESC';
  const countQuery = query.replace(
    `SELECT s.*, d.name as driver_name, d.vehicle_plate, o.order_number,\n           c.name as customer_name`,
    'SELECT COUNT(*) as count'
  );
  const total = db.prepare(countQuery).get(...params).count;
  const offset = (parseInt(page) - 1) * parseInt(limit);
  query += ` LIMIT ${limit} OFFSET ${offset}`;
  const shipments = db.prepare(query).all(...params);
  res.json({ data: shipments, total, page: parseInt(page), limit: parseInt(limit) });
});

router.get('/track/:trackingNumber', (req, res) => {
  const shipment = db.prepare(`
    SELECT s.*, d.name as driver_name, d.phone as driver_phone, d.vehicle_plate,
           o.order_number, c.name as customer_name
    FROM shipments s
    LEFT JOIN drivers d ON s.driver_id = d.id
    LEFT JOIN orders o ON s.order_id = o.id
    LEFT JOIN customers c ON o.customer_id = c.id
    WHERE s.tracking_number = ?
  `).get(req.params.trackingNumber);
  if (!shipment) return res.status(404).json({ error: 'Shipment not found' });
  const events = db.prepare('SELECT * FROM shipment_events WHERE shipment_id = ? ORDER BY timestamp DESC').all(shipment.id);
  res.json({ ...shipment, events });
});

router.get('/:id', (req, res) => {
  const shipment = db.prepare(`
    SELECT s.*, d.name as driver_name, d.phone as driver_phone, d.vehicle_plate, d.vehicle_type,
           o.order_number, c.name as customer_name, w.name as warehouse_name
    FROM shipments s
    LEFT JOIN drivers d ON s.driver_id = d.id
    LEFT JOIN orders o ON s.order_id = o.id
    LEFT JOIN customers c ON o.customer_id = c.id
    LEFT JOIN warehouses w ON s.warehouse_id = w.id
    WHERE s.id = ?
  `).get(req.params.id);
  if (!shipment) return res.status(404).json({ error: 'Shipment not found' });
  const events = db.prepare('SELECT * FROM shipment_events WHERE shipment_id = ? ORDER BY timestamp DESC').all(req.params.id);
  res.json({ ...shipment, events });
});

router.post('/', (req, res) => {
  const { order_id, driver_id, warehouse_id, origin, destination, estimated_delivery, weight, distance, cost, notes } = req.body;
  if (!origin || !destination) return res.status(400).json({ error: 'Origin and destination are required' });
  const id = uuidv4();
  const count = db.prepare('SELECT COUNT(*) as count FROM shipments').get().count;
  const trackingNumber = `TRK-${String(count + 20001).padStart(6, '0')}`;
  db.prepare(`INSERT INTO shipments (id, tracking_number, order_id, driver_id, warehouse_id, status, origin, destination, estimated_delivery, weight, distance, cost, notes) VALUES (?, ?, ?, ?, ?, 'pending', ?, ?, ?, ?, ?, ?, ?)`)
    .run(id, trackingNumber, order_id || null, driver_id || null, warehouse_id || null, origin, destination, estimated_delivery || null, weight || 0, distance || 0, cost || 0, notes || null);
  db.prepare('INSERT INTO shipment_events (id, shipment_id, event_type, location, description) VALUES (?, ?, ?, ?, ?)')
    .run(uuidv4(), id, 'created', origin, 'Shipment created and ready for pickup');
  res.status(201).json(db.prepare('SELECT * FROM shipments WHERE id = ?').get(id));
});

router.put('/:id', (req, res) => {
  const shipment = db.prepare('SELECT * FROM shipments WHERE id = ?').get(req.params.id);
  if (!shipment) return res.status(404).json({ error: 'Shipment not found' });
  const { status, driver_id, estimated_delivery, actual_delivery, notes, event_location, event_description } = req.body;
  const newStatus = status || shipment.status;
  db.prepare(`UPDATE shipments SET status=?, driver_id=?, estimated_delivery=?, actual_delivery=?, notes=?, updated_at=datetime('now') WHERE id=?`)
    .run(newStatus, driver_id ?? shipment.driver_id, estimated_delivery ?? shipment.estimated_delivery, actual_delivery ?? shipment.actual_delivery, notes ?? shipment.notes, req.params.id);
  if (status && status !== shipment.status) {
    const eventDescriptions = {
      in_transit: 'Package is in transit',
      out_for_delivery: 'Package is out for delivery',
      delivered: 'Package has been delivered',
      failed: 'Delivery attempt failed',
    };
    db.prepare('INSERT INTO shipment_events (id, shipment_id, event_type, location, description) VALUES (?, ?, ?, ?, ?)')
      .run(uuidv4(), req.params.id, status, event_location || 'Unknown', event_description || eventDescriptions[status] || `Status updated to ${status}`);
  }
  res.json(db.prepare('SELECT * FROM shipments WHERE id = ?').get(req.params.id));
});

router.delete('/:id', (req, res) => {
  const shipment = db.prepare('SELECT * FROM shipments WHERE id = ?').get(req.params.id);
  if (!shipment) return res.status(404).json({ error: 'Shipment not found' });
  db.prepare('DELETE FROM shipment_events WHERE shipment_id = ?').run(req.params.id);
  db.prepare('DELETE FROM shipments WHERE id = ?').run(req.params.id);
  res.json({ message: 'Shipment deleted successfully' });
});

module.exports = router;

File: backend/app/routes/shipments.py
import uuid
from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.database import get_db
from app.models.shipment import Shipment, ShipmentStatus, ShipmentPriority
from app.schemas.shipment import ShipmentCreate, ShipmentUpdate, ShipmentResponse

router = APIRouter(prefix="/api/shipments", tags=["shipments"])


def generate_tracking_number() -> str:
    short_id = uuid.uuid4().hex[:10].upper()
    return f"LOG-{short_id}"


@router.get("", response_model=list[ShipmentResponse])
def list_shipments(
    status: Optional[ShipmentStatus] = None,
    priority: Optional[ShipmentPriority] = None,
    search: Optional[str] = None,
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db),
):
    query = db.query(Shipment)
    if status:
        query = query.filter(Shipment.status == status)
    if priority:
        query = query.filter(Shipment.priority == priority)
    if search:
        search_term = f"%{search}%"
        query = query.filter(
            (Shipment.tracking_number.ilike(search_term))
            | (Shipment.sender_name.ilike(search_term))
            | (Shipment.recipient_name.ilike(search_term))
            | (Shipment.origin_city.ilike(search_term))
            | (Shipment.destination_city.ilike(search_term))
        )
    return query.order_by(Shipment.created_at.desc()).offset(skip).limit(limit).all()


@router.get("/count")
def count_shipments(
    status: Optional[ShipmentStatus] = None,
    db: Session = Depends(get_db),
):
    query = db.query(func.count(Shipment.id))
    if status:
        query = query.filter(Shipment.status == status)
    return {"count": query.scalar()}


@router.get("/{shipment_id}", response_model=ShipmentResponse)
def get_shipment(shipment_id: str, db: Session = Depends(get_db)):
    shipment = db.query(Shipment).filter(Shipment.id == shipment_id).first()
    if not shipment:
        raise HTTPException(status_code=404, detail="Shipment not found")
    return shipment


@router.get("/track/{tracking_number}", response_model=ShipmentResponse)
def track_shipment(tracking_number: str, db: Session = Depends(get_db)):
    shipment = db.query(Shipment).filter(Shipment.tracking_number == tracking_number).first()
    if not shipment:
        raise HTTPException(status_code=404, detail="Shipment not found")
    return shipment


@router.post("", response_model=ShipmentResponse, status_code=201)
def create_shipment(data: ShipmentCreate, db: Session = Depends(get_db)):
    shipment = Shipment(
        id=str(uuid.uuid4()),
        tracking_number=generate_tracking_number(),
        **data.model_dump(),
    )
    db.add(shipment)
    db.commit()
    db.refresh(shipment)
    return shipment


@router.patch("/{shipment_id}", response_model=ShipmentResponse)
def update_shipment(shipment_id: str, data: ShipmentUpdate, db: Session = Depends(get_db)):
    shipment = db.query(Shipment).filter(Shipment.id == shipment_id).first()
    if not shipment:
        raise HTTPException(status_code=404, detail="Shipment not found")

    update_data = data.model_dump(exclude_unset=True)
    if "status" in update_data:
        new_status = update_data["status"]
        if new_status == ShipmentStatus.IN_TRANSIT and not shipment.shipped_at:
            shipment.shipped_at = datetime.now(timezone.utc)
        elif new_status == ShipmentStatus.DELIVERED and not shipment.actual_delivery:
            shipment.actual_delivery = datetime.now(timezone.utc)

    for key, value in update_data.items():
        setattr(shipment, key, value)

    shipment.updated_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(shipment)
    return shipment


@router.delete("/{shipment_id}", status_code=204)
def delete_shipment(shipment_id: str, db: Session = Depends(get_db)):
    shipment = db.query(Shipment).filter(Shipment.id == shipment_id).first()
    if not shipment:
        raise HTTPException(status_code=404, detail="Shipment not found")
    db.delete(shipment)
    db.commit()
